ros_ws/src/controller/src/controller_node.cpp

```diff
@@ -0,0 +1,287 @@
+using namespace std::chrono_literals;
+
+ControllerNode::ControllerNode(const std::string model_path, const int num_observations, const int num_actions)
+  :Node("controller"),
+  model(model_path, num_observations, num_actions),
+  last_time(this->now()),
+  last_call_back_time(this->now()) {
+
+  // initilize control loop timer
+  timer = create_wall_timer(50ms, bind(&ControllerNode::control_loop, this));
+
+  // serial communication 
+  std::string path = "/dev/ttyTHS1";
+  serial_file_desc = open_serial(path);
+  if (serial_file_desc == -1) {
+    serial_open = false;
+  }
+  serial_open = true;
+}
+
+
+void ControllerNode::control_loop() {
+  float dt = (get_clock()->now() - last_time).seconds();
+
+  if (!serial_open) {
+    std::cout << "Serial port is not availaible\n";
+    return;
+  }
+  // get imu data
+  int read_bytes = read_serial(this->imu_data);
+  if (read_bytes == -1) return;
+  auto& input_buffer = model.get_input_buffer();
+
+  //load input buffer
+  input_buffer[0] = imu_data[0];
+  input_buffer[1] = imu_data[1];
+  input_buffer[2] = prev_actions[0];
+  input_buffer[3] = prev_actions[1];
+
+  model.run();
+  // publish motor
+  motor_cmd = { model.get_output_buffer().at(0),model.get_output_buffer().at(1) };
+  int send_bytes = write_serial(this->motor_cmd);
+
+
+  //display the results 
+  std::ostringstream out;
+  out << "Control loop :\n";
+  out << "Time step : \t" << dt << "\n";
+  out << "Data read : \t " << read_bytes << "\n";
+  out << "Roll : \t" << imu_data[0] << "\n";
+  out << "Angular velocity : \t" << imu_data[1] << "\n";
+  out << "\nINPUTS:\t";
+  for (auto element : model.get_input_buffer()) {
+    out << element << " ";
+  }
+  out << "\nOUTPUTS\t";
+  for (auto element : model.get_output_buffer()) {
+    out << element << " ";
+  }
+  out << "\nData sent : \t " << send_bytes << "\n";
+  out << std::endl;
+
+
+
+  prev_actions = { model.get_output_buffer().at(0),model.get_output_buffer().at(1) };
+  last_time = get_clock()->now();
+
+  // motor_cmd.header.stamp = get_clock()->now();
+  // motor_cmd.effort = { model.get_output_buffer().at(0), model.get_output_buffer().at(1) };
+
+  // pub->publish(motor_cmd);
+
+  RCLCPP_INFO(get_logger(), out.str().c_str());
+  std::cout << out.str() << std::endl;
+}
+
+
+/*
+ * function configures the current process to use the SCHED_FIFO scheduling policy with a specified
+ * real-time priority. It initializes the scheduling parameters, sets the priority, and applies the
+ * scheduling policy using the sched_setscheduler system call. If the priority setting fails, an error
+ * message is logged; otherwise, an informational message with the current priority is logged.
+ *
+ * Parameters:
+ * - controller: A shared pointer to the ControllerNode class, which provides the logging mechanism.
+ *
+ * Returns:
+ * - void
+*/
+
+void set_real_time_priority(std::shared_ptr<ControllerNode> controller) {
```

Comment:
This should be at the top of the file, since it isn't a class function (or even better, in a utils file).

Reply:
Yes!